3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/pii/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,301 @@
+ * identifiable information) and apply some anonymization (currently only pseudonymization) on them. Currently a single
+ * strategy for all the fields is supported due to the configuration format, and there is only one implemented strategy,
+ * however the enrichment supports a strategy per field.
+ *
+ * The user may specify two types of fields in the config `pojo` or `json`. A `pojo` field is effectively a scalar field in the
+ * EnrichedEvent, whereas a `json` is a "context" formatted field and it can either contain a single value in the case of
+ * unstruct_event, or an array in the case of derived_events and contexts.
+ *
+ * @param fieldList a list of configured PiiFields
+ * @param emitIdentificationEvent whether to emit an identification event
+ * @param strategy the pseudonymization strategy to use
+ */
+case class PiiPseudonymizerEnrichment(fieldList: List[PiiField],
+                                      emitIdentificationEvent: Boolean,
+                                      strategy: PiiStrategy)
+    extends Enrichment {
+  implicit val json4sFormats = DefaultFormats + new PiiModifiedFieldsSerializer + new PiiStrategySerializer
+  def transformer(event: EnrichedEvent): Unit = {
+    val modifiedFields: ModifiedFields = fieldList.flatMap(_.transform(event, strategy))
+    event.pii = if (modifiedFields.nonEmpty) write(PiiModifiedFields(modifiedFields, strategy)) else null
+  }
+}
+
+/**
+ * Specifies a scalar field in POJO and the strategy that should be applied to it.
+ * @param fieldMutator the field mutator where the strategy will be applied
+ */
+final case class PiiScalar(fieldMutator: Mutator) extends PiiField {
+  override def applyStrategy(fieldValue: String, strategy: PiiStrategy): (String, ModifiedFields) =
+    if (fieldValue != null) {
+      val modifiedValue = strategy.scramble(fieldValue)
+      (modifiedValue, List(ScalarModifiedField(fieldMutator.fieldName, fieldValue, modifiedValue)))
+    } else (null, List())
+}
+
+/**
+ * Specifies a strategy to use, a field mutator where the JSON can be found in the EnrichedEvent POJO, a schema criterion to
+ * discriminate which contexts to apply this strategy to, and a JSON path within the contexts where this strategy will
+ * be applied (the path may correspond to multiple fields).
+ *
+ * @param fieldMutator the field mutator for the JSON field
+ * @param schemaCriterion the schema for which the strategy will be applied
+ * @param jsonPath the path where the strategy will be applied
+ */
+final case class PiiJson(fieldMutator: Mutator, schemaCriterion: SchemaCriterion, jsonPath: String) extends PiiField {
+  implicit val json4sFormats = DefaultFormats
+
+  override def applyStrategy(fieldValue: String, strategy: PiiStrategy): (String, ModifiedFields) = {
+    val modifiedFields = MutableList[JsonModifiedField]()
+    if (fieldValue != null) {
+      (compact(render(parse(fieldValue) match {
```

Comment:
I think @BenFradet already raised that it's very hard to read this lisp-ish notation. Can we split this into separate values (and preferably smaller functions).